### PR TITLE
gh-116402: Avoid readline in test_builtin TTY input tests

### PR DESCRIPTION
**Pull request by ambv**
The intention here is to make the TTY `input()` tests run in as many scenarios as possible. For this to happen, we need to temporarily remove the function pointer set to `call_readline` in `PyInit_readline()`. This pointer is normally `NULL` without `readline` and `PyOS_Readline` uses a basic implementation of its own in this case. This is what we want.

We could try and avoid importing `readline` in tests, but we have played this whack-a-mole for a long time. For instance, now that `pdb` uses `rlcompleter`, `test_builtin` imports that always (since it imports `doctest` in `load_tests()` and `doctest` imports `pdb` as it subclasses it). I think we should stop trying to not load `readline`, and just detach its function temporarily. This avoids discussions of "unimporting" readline, which would be problematic as `PyInit_readline()` sets global state on the C side through `setup_readline`.

While CI is famously not run with a TTY and those tests will be skipped there, people running tests sequentially from their terminal will now be able to have those tests run in many more situations than before.

I did the function pointer unsetting with `ctypes` for simplicity. If that's unavailable, the context manager simply skips the tests.

* Issue: gh-116402

### CODE REVIEW

**Comment by ambv** [comment]
@Eclips4, can you confirm this fixes the issue for you?

**Comment by Eclips4** [comment]
> @Eclips4, can you confirm this fixes the issue for you?

Yes! Thanks for the fix!

File: Lib/test/test_builtin.py [hunk 1]
```diff
@@ -2400,28 +2403,34 @@ def child(wpipe):
...
         # the readline module, but the builtins module.
-        if 'readline' in sys.modules:
-            self.skipTest("the readline module is loaded")
+        if "readline" in sys.modules:
+            c = import_module("ctypes")
+            fp_api = "PyOS_ReadlineFunctionPointer"
+            prev_value = c.c_void_p.in_dll(c.pythonapi, fp_api).value
+            c.c_void_p.in_dll(c.pythonapi, fp_api).value = None
+            yield
+            c.c_void_p.in_dll(c.pythonapi, fp_api).value = prev_value
```
Comment (JelleZijlstra) [inline Lib/test/test_builtin.py]: Put this in a finally block?

File: Lib/test/test_builtin.py [hunk 2]
```diff
@@ -2400,28 +2403,34 @@ def child(wpipe):
...
-        # non-ASCII characters. Skip tests on non-ASCII characters if the
-        # readline module is loaded, since test_builtin is not intended to test
+        # non-ASCII characters.
+        # Unlink readline temporarily from PyOS_Readline() for those tests,
+        # since test_builtin is not intended to test
         # the readline module, but the builtins module.
-        if 'readline' in sys.modules:
-            self.skipTest("the readline module is loaded")
+        if "readline" in sys.modules:
+            c = import_module("ctypes")
```
Comment (JelleZijlstra) [inline Lib/test/test_builtin.py:2441]: Why not just `import ctypes`?

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, this also fixes `test_builtin` for me locally.

File: Lib/test/test_builtin.py (hunk 2 above)
Comment (ambv) [reply Lib/test/test_builtin.py:2441]: Because ctypes is technically an optional module.

Comment (JelleZijlstra) [reply Lib/test/test_builtin.py:2441]: If it's not present line 2418 below will error out anyway. You could do `import ctypes` right here so the global scope doesn't rely on ctypes being present.

Comment (ambv) [reply Lib/test/test_builtin.py:2441]: `test.support.import_helper.import_module` raises `unittest.SkipTest` if the import doesn't work so there will be no failure on 2418 as the execution won't reach there.

File: Lib/test/test_builtin.py (hunk 1 above)
Comment (ambv) [reply Lib/test/test_builtin.py]: I'll wait for the buildbots to finish and push the `try:finally:`

**Comment by ambv** [comment]
All buildbots (including refleaks and bigmem) passed save for "iOS ARM64 Simulator" with an unrelated failure. Rebasing on main with the `try:finally:` suggestion and landing as soon as CI passes.

**Pull request merged**